### DISCUSSION TITLE
Feature Request: Update persist middleware so setItem is only called when data being persisted has changed

### DISCUSSION THREAD

**Discussion by denno020** [pmndrs/zustand#1135] [Ideas]
**Background**
Since the `partialize` function has been added to the `persist` middleware, we've been able to customize the data that is passed to the `setItem` function within `getStorage`. This also gives us the ability to remove certain properties that are within a Zustand store so they're not persisted.

In practice, it appears that `setItem` will be called for every single state update, even if the piece of state being updated isn't being persisted.

**Demo**
Please see this very contrived example of a store that contains a user property. This could be some form of authentication information that we don't want to persist, so we exclude using the `partialize` function. However, every time the value of `name` changes, the `setItem` function will be called, even though the data is exactly the same each time
https://codesandbox.io/s/long-resonance-46hry8

**Request**
Ideally this `setItem` function would only be called when the data that is being persisted has actually changed.

**Why**
If we're using persist to store to localStorage, then it probably doesn't matter. However, as soon as we introduce some sort of async storage, perhaps an online database through a cloud function, extra calls to the database that don't actually update anything could have cost implications in the app's hosting environment. As the app scales to many users, there could be many more calls to update the database that aren't necessary, and therefore the costs for the app hosting could skyrocket.

**How**
`setItem` is called with a `value` that is simply a string. We could store the previous value in memory and do a simple comparison to see if the data has changed, only triggering `setItem` when that data has changed.
I'm also open to any other ideas, this is just the most simple that I thought would fit in with Zustand's lightweight approach.

**Alternative**
If we don't want to implement this into Zustand core, I think it would be really beneficial to at least give a mechanism, either through new middleware or a new function within `persist`, to be able to handle such comparison ourselves

**Comment by dai-shi** [comment]
Sounds like a controversial one. Someone may want to avoid double calling `serialize`. And, someone may want to avoid extra comparison (especially if they don't use `partialize`. This seems depending on various requirements. I feel like this should be handled optionally. What you can do now is to do the comparison at the storage level. We could create a wrapper util like `getStorage: () => wrapper(yourStorage)`. Not to add it to core, maybe a recipe in docs is fine.

**Comment by denno020** [comment]
I'm not sure it's too controversial. Intuitively, I wouldn't expect `setItem` to be called when the arguments are exactly the same as the immediately previously called time.

As we can see in the codesandbox demo, every type of the keyboard is calling `setItem` with an identical value. There's nothing changing, so there's no need to call `setItem`.

Personally I've worked around this in my app with a basic variable that keeps track of the previous argument:

```
import create from "zustand";
import { persist } from "zustand/middleware";

let prevPersistValue = '';

const useStore = create(
  persist(
    (set, get) => ({/* ... */}),
    {
      name: "app-state",
      getStorage: () => ({
        // ...
        setItem: (name, value) => {
          if (prevPersistValue === value) return;

          prevPersistValue = value;
          // Handle setting the data
        }
      })
    }
  )
);
export default useStore;
```

But that seems like something we shouldn't have to do in our apps, because Zustand has all of that same information available, so should be able to hold back on calling `setItem` when it's not necessary

**Comment by dai-shi** [comment]
Well, it seems to me that it's totally fine to do it in apps. The library don't know about how `serialize` and `partialize` are implemented. So, dealing for all cases can be complicated (hmm, now I think we shouldn't make `serialize` async-able...), and lead to overhead for users who don't use `serialize` and `partialize`.

**Comment by denno020** [comment]
Oh maybe I'm missing a piece of the puzzle.. Is `serialize` used to control how `value` looks when it's passed to `setItem`? If that's the case then I understand the hesitation as Zustand won't know the implementation and therefore how the data looks. In that case, it would be great if there was some way of returning (or not returning) data from `serialize` that would allow us to control if `setItem` is called.. But maybe that's getting a bit too complicated, perhaps we do just need something added to docs, as you initially mentioned

**Comment by dai-shi** [comment]
Yeah, probably.

Another nit is `prevPersistValue` can be outdated compared to the stored value.

One idea is, as this is the issue with `partialize`, if `partialize` returns a same reference as before, we can bail out. Then, people can use memoization library for `partialize` to avoid double invoking.

**Comment by denno020** [comment]
Oh I don't think we should worry about what is actually in the persisted store already. That would introduce quite a bit more overhead if we were comparing the value about to be saved against the value that is already saved. I'm more thinking of keeping it super simple: we can remember the most recently passed arguments, and if they're the same, don't call `setItem` again. If the user happens to refresh the browser, and therefore blow away any values stored in memory, then I think it's acceptable to have a potentially unnecessary call to `setItem`, because as far as the app is concerned, it hasn't called `setItem` yet.

In the [discussion](https://github.com/pmndrs/zustand/discussions/893) that I opened on the topic, I mentioned memoization, but when I saw that `value` being passed to `setItem` was only a string, didn't think we would need a memoization library for that. However if users are able to send through objects to `setItem`, then having a way to implement memoization would be great

**Comment by dai-shi** [comment]
I think we are on the same page. My hesitation is, while `setItem` is string based, it's just because we have `serialize` and we don't know how costly it is. So, from the library perspective, comparing strings isn't ideal because it comes with the assumption that `serialize` is lightweight. My preference is object based solution.

**Reply by 117** [reply]
serialize is deprecated

**Reply by dai-shi** [reply]
Great point.

https://github.com/pmndrs/zustand/discussions/1135#discussioncomment-3226897
I'm okay to make it so that if `partialize` returns the same reference, it bails out. But this is tricky, because we pass a copy to `partialize`. I don't think anybody would get a benefit from it.

**Comment by dai-shi** [comment]
While I understand there's room for improvement, let's hold this on.
We should collect more feedbacks on problems and use cases.

**Comment by timoisalive** [comment]
This is an interesting topic. In my case, I have a store with data which gets updated seldom and data which is updated very frequently. Only the data which is updated seldom is persisted, but unfortunately the `setItem` gets called on every update to the store, meaning very frequently.

How should I be able to tell in `setItem` which data has changed, in order to prevent the unnecessary persisting?

**Reply by timoisalive** [reply]
I didn't come up with a true solution yet, but a workaround to mitigate the performance issue in my case:

I'm debouncing the `setItem` function, so it doesn't get run on each update:
```
  setItem: debounce(
    async (
      key: string,
      value: StorageValue<Partial<MyStoreState>>
    ) => {
      await MMKVStorage.setItem(key, JSON.stringify(value))
    },
    250
  ),
```

**Comment by 117** [comment]
bump

**Reply by dai-shi** [reply]
So, what's your use case? Or, you are interested in the feature in general.

Would you like to to try one approach? https://github.com/pmndrs/zustand/discussions/1135#discussioncomment-7181070
